File: src/main/java/chapter3/ReduceExample.java
package chapter3;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.functions.BiFunction;


public class ReduceExample {

    public void marbleDiagram() {
        String[] balls = {"1", "3", "5"};
        Maybe<String> source = Observable.fromArray(balls)
                .reduce((ball1, ball2) -> ball2 + "(" + ball1 + ")");
        source.subscribe(System.out::println);
    }

    public void biFunction() {
        BiFunction<String, String, String> mergeBalls =
                (ball1, ball2) -> ball2 + "(" + ball1 + ")";
        String balls[] = {"1", "3", "5"};
        Maybe<String> source = Observable.fromArray(balls)
                .reduce(mergeBalls);
        source.subscribe(System.out::println);
    }

    public static void main(String[] args) {
        ReduceExample demo = new ReduceExample();
        demo.marbleDiagram();
        demo.biFunction();
    }

}

File: src/main/kotlin/chapter3/ReduceExample_.kt
package chapter3

import io.reactivex.Observable


class ReduceExample_ {

    fun marbleDiagram() {
        val balls = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*balls)
                .reduce { ball1, ball2 -> "$ball2($ball1)" }
        source.subscribe { println(it) }
    }

    fun biFunction() {
        val mergeBalls = { ball1: String, ball2: String -> "$ball2($ball1)" }
        val balls = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*balls)
                .reduce { t1, t2 -> mergeBalls(t1, t2) }
        source.subscribe { println(it) }
    }

}


fun main(args: Array<String>) {
    val demo = ReduceExample_()
    demo.marbleDiagram()
    demo.biFunction()
}